modules/dop/providers/projectpipeline/service_impl.go

```diff
@@ -720,22 +720,29 @@ func (p *ProjectPipelineService) Cancel(ctx context.Context, params deftype.Proj
 		return nil, err
 	}
 
-	runningPipelineID, err := p.getRunningPipeline(extraValue.CreateRequest.PipelineSource.String(), extraValue.CreateRequest.PipelineYmlName)
+	pipelineInfo, err := p.bundle.GetPipeline(uint64(definition.PipelineId))
 	if err != nil {
 		return nil, apierrors.ErrCancelProjectPipeline.InternalError(err)
 	}
-	if runningPipelineID == 0 {
-		return nil, apierrors.ErrCancelProjectPipeline.InternalError(fmt.Errorf("not find running pipeline, can not cancel"))
-	}
 
-	var req apistructs.PipelineCancelRequest
-	req.PipelineID = runningPipelineID
-	req.IdentityInfo = params.IdentityInfo
-	err = p.bundle.CancelPipeline(req)
-	if err != nil {
-		return nil, apierrors.ErrCancelProjectPipeline.InternalError(err)
+	if pipelineInfo.Status.IsRunningStatus() {
+		var req apistructs.PipelineCancelRequest
+		req.PipelineID = uint64(definition.PipelineId)
+		req.IdentityInfo = params.IdentityInfo
+		err = p.bundle.CancelPipeline(req)
+		if err != nil {
+			return nil, apierrors.ErrCancelProjectPipeline.InternalError(err)
+		}
+
+		_, err = p.PipelineDefinition.Update(context.Background(), &dpb.PipelineDefinitionUpdateRequest{PipelineDefinitionID: definition.ID, Status: string(apistructs.PipelineStatusStopByUser), PipelineId: definition.PipelineId})
```

Comment:
maybe you can do this inside cancelPipeline?

Reply:
Interfaces such as execution, execution at repeated failures, Retry, cancel execution, and start timing should be considered in the pipeline component, which should be reconstructed with another pr